zend/callable.h

```diff
@@ -198,6 +199,24 @@ class Callable
             case Type::Object:      info->type_hint = IS_OBJECT;    break;  // must be an object of the given classname
             case Type::Callable:    info->type_hint = IS_CALLABLE;  break;  // anything that can be invoked
             default:                info->type_hint = IS_UNDEF;     break;  // if not specified we allow anything
+#elif PHP_VERSION_ID >= 80000
+
+            case Type::Undefined:   info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_UNDEF, 0, 0);     break;  // undefined means we'll accept any type
+            case Type::Null:        info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_UNDEF, (int)arg.allowNull(), 0);     break;  // this is likely an error, what good would accepting NULL be? accept anything
+            case Type::False:       info->type = (zend_type) ZEND_TYPE_INIT_CODE(_IS_BOOL, (int)arg.allowNull(), 0);     break;  // accept true as well ;)
+            case Type::True:        info->type = (zend_type) ZEND_TYPE_INIT_CODE(_IS_BOOL, (int)arg.allowNull(), 0);     break;  // accept false as well
+            case Type::Bool:        info->type = (zend_type) ZEND_TYPE_INIT_CODE(_IS_BOOL, (int)arg.allowNull(), 0);     break;  // any bool will do, true, false, the options are limitless
+            case Type::Numeric:     info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_LONG, (int)arg.allowNull(), 0);      break;  // accept integers here
+            case Type::Float:       info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_DOUBLE, (int)arg.allowNull(), 0);    break;  // floating-point values welcome too
+            case Type::String:      info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_STRING, (int)arg.allowNull(), 0);    break;  // accept strings, should auto-cast objects with __toString as well
+            case Type::Array:       info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_ARRAY, arg.allowNull(), 0);     break;  // array of anything (individual members cannot be restricted)
+            case Type::Object:                                                            // if there is a classname and the argument is not nullable, it's simply the classname
+                if (!arg.classname()) info->type = (zend_type) ZEND_TYPE_INIT_CODE(IS_OBJECT, arg.allowNull(), 0);
```

Comment:
All of these gcc 7 complain about narrowing of int to uint32_t, I figured out its because of the arg.allowNull(), but I can't make the warning go away - I don't think the warning is a big issue but its nice not to get a warning on every file compilation

Reply:
These are errors on clang 11, so it will be an issue going forward. The problem is because the zend_type struct takes a `uint32_t`, but they defined `_ZEND_TYPE_NULLABLE_BIT` as `0x2`, and in C/C++ constants are signed by default. I'm not sure this can be solved in a clean way without getting into their code, but I was able to undef annd redef the macro as `0x2u` and the issue is gone.

Reply:
so I'm fairly ignorant of everything going on in this part, but how come when I pass a 0 (completely ignoring the arg.allowNull() call) it compiles? How is that related to the _ZEND_TYPE_NULLABLE_BIT? Is there no way to cast what we're providing to the ZEND_TYPE_INIT_CODE so that its the proper size to begin with?

Reply:
I filed a bug upstream:

https://bugs.php.net/bug.php?id=80617

Reply:
They've committed a fix so I expect this issue to go away in 8.0.2